jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/transformation/FoldingTransformer.java

```diff
@@ -52,274 +57,83 @@
-        for (V v : vertices)
-        {
-            newGraph.addVertex(v);
-            for (V s : g.getSuccessors(v))
-            {
-                for (V t : g.getSuccessors(s))
-                {
-                    if (!vertices.contains(t) || t.equals(v)) 
-                        continue;
-                    newGraph.addVertex(t);
-                    newGraph.addEdge(edge_factory.get(), v, t);
-                }
-            }
+    	Preconditions.checkArgument(g.nodes().containsAll(nodes),
+    			"Input graph must contain all specified nodes");
+        MutableGraph<N> newGraph = GraphBuilder.from(g).expectedNodeCount(nodes.size()).build();
+
+        for (N node : nodes) {
+        	for (N s : g.successors(node)) {
+        		for (N t : g.successors(s)) {
+        			if (!nodes.contains(t) || t.equals(node)) {
+        				continue;
+        			}
+        			newGraph.putEdge(node, t);
+        		}
+        	}
         }
         return newGraph;
     }
 
     /**
-     * Converts <code>g</code> into a unipartite graph whose vertices are the
-     * vertices of <code>g</code>'s partition <code>p</code>, and whose edges
-     * consist of collections of the intermediate vertices from other partitions.  
-     * For vertices
+     * Converts <code>g</code> into a unipartite graph whose vertex set is the
+     * vertices of <code>g</code>'s partition <code>p</code>.  For vertices
      * <code>a</code> and <code>b</code> in this partition, the resultant
      * graph will include the edge <code>(a,b)</code> if the original graph
      * contains edges <code>(a,c)</code> and <code>(c,b)</code> for at least
      * one vertex <code>c</code>.
      * 
      * <p>The vertices of the new graph are the same as the vertices of the
-     * appropriate partition in the old graph; the edges in the new graph are
-     * collections of the intermediate vertices <code>c</code>.
+     * appropriate partition in the input graph.  The edge values are the sets of
+     * nodes that connected the edge's endpoints in the input graph.
      * 
      * <p>This function will not create self-loops.
      * 
-     * @param <V> vertex type
+     * @param <N> vertex type
      * @param <E> input edge type
-     * @param g input k-partite graph
-     * @param p predicate specifying vertex partition
-     * @param graph_factory Supplier used to create the output graph 
-     * @return the result of folding g into unipartite graph whose vertices
-     * are those of the <code>p</code> partition of g
+     * @param g input graph
+     * @param nodes input node set
      */
-    public static <V,E> Graph<V, Collection<V>> foldKPartiteGraph(KPartiteGraph<V,E> g, Predicate<V> p, 
-            Supplier<Graph<V, Collection<V>>> graph_factory)
+	// TODO: consider providing ValueGraph/Network versions of this
+	// TODO: consider renaming this
+    public static <N> ValueGraph<N, Set<N>> foldToValueGraph(Graph<N> g, Set<N> nodes)
```

Comment:
@jrtom I think the return type should be either `MutableValueGraph<N, Set<N>` or `ValueGraph<N, ImmutableSet<N>>`, so that the mutability of the entire data structure is kept consistent.